rfcs/007-background-workers.md

```diff
@@ -0,0 +1,252 @@
+
+When a job finishes running, either because it was a successful completion or a
+terminal failure, the worker MUST remove the job from its active queue and
+clear its `WorkerId` field. Then, based on type-specific behavior, it MUST do
+either one of:
+
+ - Update the `JobState` to `"completed"` and leave the job in the jobs table.
+   It would be up to the job creation logic to inspect and remove it.
+ - Delete the job from the system altogether. This may be useful when running
+   _replicate jobs, for example.
+
+
+If a user decided to cancel a job that is running, they MUST toggle `CancelReq`
+to `true`. Then they MUST wait for the job to stop running. If the job is not
+running, they may directly remove the job from the pending and the main jobs
+table.
+
+
+## Worker Lifecyle
+
+Workers can be dynamically started and stopped. When workers start, they MUST
+register themselves in the `"workers"` subspace based on their `WorkerType`.
+Each change to the workers list, should be accompanied by updating the
+`"workers_vs"` versionstamp. That versionstamp MAY be efficiently monitored for
+changes using a watch.
+
+
+### Worker Liveliness Monitoring
+
+As long as the workers are active and ready to process jobs, they MUST
+periodically ping their health key. The period MUST be less than the
+`WorkerTimeout` value. Other worker nodes, for example, one of the "neighbors"
+(to left or to the right) of the worker, will monitor that health entry. If a
+worker dies unexpectedly, the neighbor MUST remove all of the dead worker's
+jobs from the active queue and submit them back into the pending queue.
+
+
+### Liveliness Race Condition
+
+In case workers come back alive after not updating their health entry often
+enough, and another worker had cleared and re-submitted they jobs, workers MUST
+monitor their active queue and if they detect that it was cleared or `WorkerId`
+was updated and they are no longer the current `WorkerId`, they MUST stop
+running the job and go back to the pending queue to pick new jobs to run.
+
+### Clean Worker Shutdown
+
+If a worker is stopped cleanly, they MUST stop accepting new jobs from the
+pending queue, stop running their jobs, and then resubmit those jobs back into
+the pending queue. Finally, they MUST clear their health entry and remove
+themselves from workers membership list.
+
+### User Job Cancellation
+
+If a user decided to a cancel an active job. They MUST update the `CancelReq`
+field to `true`. Workers should periodically monitor that field and if they
+field is true, they should stop running the job, update its jobs state and
+remove it from their active queue.
+
+## `/_active_tasks` API Compatibility
+
+The shape of the active queue was deliberately constructed with the hopes of
+using it to implement the `/_active_tasks` API endpoint.
+
+When the HTTP API is access, a single `get_range(...)` request for
+`("couch_workers", "active")` subspace MAY be used to generate the results.
```

Comment:
Yeah that is a good idea. I couldn't think of a good shape for the API yet and left it out. Once we get a feel for it, (play with a prototype), it might be easier to see what it might look like.